### ISSUE TITLE
Table cell child elements have "visible" attribute set to false

### ISSUE THREAD

**Issue by danail-branekov** [appium/appium#2582]
Hi,

I have created a simple IOS application with a table which is being populated upon clicking a button. The newly created table cells and their content is visible on the device. They are also available in the Appium inspector model but their "visible" attribute is set to "false"

I created the following script for Apple instruments for the very same application:
```
var target = UIATarget.localTarget();

var mainWindow = target.frontMostApp().mainWindow();
mainWindow.buttons()["Add New Cell"].tap();

var tableView = mainWindow.tableViews()["Empty list"];

var tableCells = tableView.cells();
for(var cellIdx = 0; cellIdx < tableCells.length; cellIdx++) {
	var cellElements = tableCells[cellIdx].elements();
	for(var elementIdx = 0; elementIdx < cellElements.length; elementIdx++) {
		var currentElement = cellElements[elementIdx];
		UIALogger.logMessage("Element with name " + currentElement.name() + "; visible=" + currentElement.isVisible());
	}
}

UIALogger.logMessage("Done")

```
The script would press the button for adding a new table cell and would traverse the cell elements printing their "visible" value. In the log I get the following results:
```
Element with name Text1; visible=1
Element with name Text 2; visible=1
Element with name null; visible=1
```
As you can see according to instruments all the elements are visible. However, Appium would return visible=false
I would also like to mention that when you add more rows in the table (i.e. more instances of UIATableCell) the visible flag is sometimes true. I cannot find a stable scenario when this happens

The sample application can be found at https://github.com/danail-branekov/samples/tree/tableviewcell-visible-bug

Here is my environment
* OSX 10.9.2
* XCode/Instruments 5.1.1
* Appium 0.17.6

Thanks!

Regards, Danail

**Comment by ankur-sharma2014** [comment]
Is this issue resolved?  I am using 1.3.4 on iOS 7.1 and still get all the elements 'visibility' showing up as 'false'. I checked the source code, its fine there.

**Comment by bootstraponline** [comment]
Is it any different on iOS 8? I think this problem is app specific and related to the version of iOS/instruments being used.

**Comment by danail-branekov** [comment]
I just built my sample application for IOS8.1 and gave it a try with Instruments 6.1. Yes, the results are the same. The instruments script returns "1" when checking whether the controls in a table cell is visible. Here is what Appium application XML looks like (excerpt for the table cell only)
``` xml
<UIATableCell name="Text1" label="" value="" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/1/1/1" x="23.4375" y="87.40625" width="328.125" height="51.5625">
    <UIATextField name="Text1" label="Text1" value="Text 1: 1" dom="" enabled="true" valid="true" visible="false" hint="" path="/0/1/1/1/0" x="46.875" y="87.40625" width="60.9375" height="24.609375">
        <UIATextField name="Text1" label="Text1" value="Text 1: 1" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/1/1/1/0/0" x="46.875" y="87.40625" width="60.9375" height="24.609375">
        </UIATextField>
    </UIATextField>
    <UIATextField name="Text 2" label="Text 2" value="Text 2: 1" dom="" enabled="true" valid="true" visible="false" hint="" path="/0/1/1/1/1" x="46.875" y="110.84375" width="60.9375" height="26.953125">
        <UIATextField name="Text 2" label="Text 2" value="Text 2: 1" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/1/1/1/1/0" x="46.875" y="110.84375" width="60.9375" height="26.953125">
        </UIATextField>
    </UIATextField>
    <UIATextField name="" label="" value="Text 3: 1/1" dom="" enabled="true" valid="true" visible="false" hint="" path="/0/1/1/1/2" x="117.1875" y="87.40625" width="219.140625" height="50.390625">
        <UIATextField name="" label="" value="Text 3: 1/1" dom="" enabled="true" valid="true" visible="true" hint="" path="/0/1/1/1/2/0" x="117.1875" y="87.40625" width="219.140625" height="50.390625">
        </UIATextField>
    </UIATextField>
</UIATableCell>

```
It is interesting to mention that each text field is represented by two (parent and child) UIATextField elements. The parent has the "visible" property set to false while the child "visible" property is true. 

Nevertheless I would not agree that this issue is application specific. After all, Instruments clearly reports that the items in the table cell are visible (returns "1"). I have also witnessed other applications with table views where sometimes the content of the first table cell is reported as invisible by Appium.

**Comment by bootstraponline** [comment]
This is probably an instruments bug that appium can work around. I think you're right that it's not app specific.

**Comment by ankur-sharma2014** [comment]
Thanks Daniel. I do not have that scenario "that each text field is represented by two (parent and child) UIATextField elements.". For my app, all elements - tablecells, textfields, buttons, 'visible' attribute is false with ios7 and 8. Instruments clearly reports that the items in the table cell are visible (returns "1"). Will be interesting to know, what was the issue that Appium was unable to recognize it is as Visible and anything we could have done better in the app.

**Comment by steher** [comment]
I have the same issue using Appium v 1.4.0.
Is there, by any chance a fix under way?

**Comment by Jonahss** [comment]
No fix scheduled as of yet, sorry.
Have you tried iOS 8.3 with Xcode 6.3.2? Apple fixed up instruments a bit in that release.

**Comment by steher** [comment]
Thanks for the response. I have tried iOS 8.3 and Xcode 6.3.2, but it's not fixed in that release

**Comment by ChiWhiteSox56** [comment]
iOS 8.3 Simulator, Appium 1.4.8

I believe I'm seeing the same issue for a UIATextField that is programatically drawn, according to the developer. 

According to the Appium inspector:
enabled = true
visible = false
valid = true

As you can see from the image attached, the text field in question is visible. My tests constantly fail after 60 seconds because this particular element is not being found. (I am using xpath and I know that my path is correct.) This text field is nested as follows: UIAApplication -> UIAWindow -> UIANavigationBar -> UIATextField

![screen shot 2015-07-27 at 4 18 37 pm](https://cloud.githubusercontent.com/assets/1609258/8918051/5aa9ac2c-347b-11e5-95e5-b6cc33e6e4dd.png)

**Comment by steher** [comment]
Don't know if it's the same issue, but these links helped me solving it:
https://discuss.appium.io/t/appium-doesnt-detect-cells-in-tableview/698/4
https://developer.yahoo.com/blogs/ydn/defining-isaccessibilityelement-ios-application-53755.html

**Comment by iAmNaz** [comment]
@steher a HUGE thank you for you. The second link "Defining isAccessibilityElement In Your iOS Application" helped me addressed the issue where the tableView in my iOS app can't be seen by the Appium Inspector. Thank you!!!
Since the VCs are embedded in a NavCtrlr I have to turn off the NavCtrlr view accessibility so that the tableview becomes visible in the inspector.

**Comment by steher** [comment]
You're welcome. I'm glad it helped you out

**Comment by SrinivasanTarget** [comment]
I think this has been fixed in xcode 7.3

**Comment by SrinivasanTarget** [comment]
@triager please close
